cluster-autoscaler/observers/scale_state_observer.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package observers
```

Comment:
I think there is value in having `observers` under the same package because it makes them easy to identify (just like `processors` directory). Maybe something like this makes more sense?
```
├── observers
│   └── nodegroupchange
│       └── scale_state_observer.go
``` 
This makes the import path as `../cluster-autoscaler/observers/nodegroupchange`. We'd still have a specific package called `nodegroupchange` while letting the user know that it's an observer. 

If we go with specific observer -> specific use case -> specific package approach, the name of the package is going to be `nodegroupchangeobserver` which IMO is a bit too long. We can shorten it to `ngchangeobserver` or `ngcobserver` but that means you need to know `ng` stands for `nodegroup` or (`nodegroupchange` if we shorten it to `ngc`). 

Open to other better approaches as well (I am leaning more towards first). Let me know what you think.

Reply:
Grouping under the same `observers` directory, but having separate packages sounds acceptable to me. I disagree we need to have `observer` as part of the package name though: referring to the object externally as `nodegroupchange.Observer` should be well readable, there's no need to repeat the same word in the package name and in the type names within the package.